### ISSUE TITLE
After instrument with dyninst binary exports functions from libc

### ISSUE THREAD

**Issue by alvarofe** [dyninst/dyninst#529]
I was testing out dyninst more specifically with version 10. I took as an example afl-dyninst to get acquaintance with the API and learn how it works - implementing minimal logic to instrument and print random id for each basic block visited. However, after instrumenting `/bin/ls` I noticed an infinite loop with the binary instrumented.

After a bit of researching I noticed that the binary instrumented somehow is exporting libc functions - `ls-inst` is `/bin/ls` instrumented.

```
ryzen:dyninst-fun alvaro$ rabin2 -E ls|grep -E "(malloc|free|__cxa_finalize)"
125 0x00016310 0x00016310 GLOBAL   FUNC  101 _obstack_free

ryzen:dyninst-fun alvaro$ rabin2 -E ls-inst|grep -E "(malloc|free|__cxa_finalize)"
125 0x00017310 0x00017310 GLOBAL   FUNC  101 _obstack_free
137 0x000056d0 0x000056d0 GLOBAL   FUNC    0 free
138 0x000056d8 0x000056d8 GLOBAL   FUNC    0 malloc
139 0x000056e0 0x000056e0 GLOBAL   FUNC    0 __cxa_finalize
345 0x00077b1e 0x00155b1e GLOBAL   FUNC 2493 _obstack_free_dyninst
```
Those exported symbols points to the `.plt.got` section therefore there will be an infinite loop when the dynamic loader resolve the relocs.

```
ryzen:dyninst-fun alvaro$ r2 ls-inst
 -- Thanks for using radare2!
[0x00007130]> s 0x000056d8
[0x000056d8]> pd 2
            ;-- malloc:
            0x000056d8      ff250ad90100   jmp qword [reloc.malloc_232] ; [0x22fe8:8]=0
            0x000056de      6690           nop
[0x000056d8]> pxw 20 @0x22fe8
0x00022fe8  0x00000000 0x00000000 0x00000000 0x00000000  ................
0x00022ff8  0x00000000 
[0x000056d8]> iS~.plt.got
13 0x000056d0    24 0x000056d0    24 -r-x .plt.got
```
The addr `0x00022fe8` will be patched by the dynamic loader with the value of `0x56d8` leading to the infinite loop. 

If I patch those things manually with gdb by the actual values, everything works as expected. 
Do you know which might be the reason of this? I don't know if I am missing something or might be a bug with dyninst. 

Attached is the playground for dyninst which I was using - rename the files from .png to .cc.
![playground](https://user-images.githubusercontent.com/3474042/51204042-f30f3580-1902-11e9-9837-a9590ffebe81.png)
![libinstdyninst](https://user-images.githubusercontent.com/3474042/51204061-faceda00-1902-11e9-9653-389e05f91e6c.png)

**Comment by mxz297** [comment]
Hi,

I just downloaded your example code and it runs fine on my machine:

```
[xmeng@coriander] (56)$ ./playground -i /bin/ls -o ls-inst
playground dyninst

Instrumenting module: ls
.....
[xmeng@coriander] (57)$ ./ls-inst 
basic block with id 13410
basic block with id 42215
basic block with id 12187
basic block with id 47478
.....
```

My machine is x86-64, running ubuntu 16. What's your machine's architecture, OS distribution, and version?

**Comment by alvarofe** [comment]
My machine is debian x86_64 and I compiled myself dyninst.

```
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Release ..
make
make install
```

```
ryzen:dyninst-fun alvaro$ apt list --installed|grep libboost-all-dev
libboost-all-dev/now 1.67.0.1 amd64 [installed,local]
ryzen:dyninst-fun alvaro$  export DYNINSTAPI_RT_LIB=/usr/local/lib/libdyninstAPI_RT.so
ryzen:dyninst-fun alvaro$ ./playground -i ls -o ls-inst
...
Basic Block 0x7a5d of targ7990 with size 5 with random id 59202                                                                                                                                                                                                                                                               
Saving the instrumented binary to ls-inst ...                                                                                                                                                                                                                                                                                 
Did a total of 4288 basic block insertions   
```
But when executing I get the infinite loop

**Comment by alvarofe** [comment]
I tried to clean up my system with unstable versions with stables ones (libboost and such), however the issue persist on `ls` which was my binary test. For instance, I tried with cat and it did work out well.

```
debian:dyninst-fun alvaro$ ./cat-inst file                                                                                                                                                                                                                                                                                    
basic block with id 48391                                                                                                                                                                                                                                                                                                     
basic block with id 46109                                                                                                                                                                                                                                                                                                     
basic block with id 18961                                                                                                                                                                                                                                                                                                     
basic block with id 35525                                                                                                                                                                                                                                                                                                     
basic block with id 44835                                                                                                                                                                                                                                                                                                     
basic block with id 16139                                                                                                                                                                                                                                                                                                     
basic block with id 20811                                                                                                                                                                                                                                                                                                     
basic block with id 41955                                                                                                                                                                                                                                                                                                     
basic block with id 36950                                                                                                                                                                                                                                                                                                     
basic block with id 33523     
```

So there must be something special with that `ls`. Attached in case you wanna test with your installation.


![ls](https://user-images.githubusercontent.com/3474042/51330539-0a236400-1a78-11e9-80c4-d05f533e781a.png)

**Comment by mxz297** [comment]
I am able to reproduce this infinite loop by using the ```ls``` binary you provided.

Your OS is debian, which seems to, by default, have position independent executable. My OS is ubuntu which does not install PIEs. So, my first guess is that Dyninst has some problem rewriting PIE.

Thanks for reporting this problem and I will get back to you later when I fixed it.

**Comment by alvarofe** [comment]
`cat` is also PIE on my machine so dunno if that will be the reason honestly  - although talking completely without knowledge about the framework. My guess is that it is related to some extent to how relocations are handled just by seeing the differences between the two binaries.

```
cat
[0x00001820]> ii~malloc
  37 0x00001820  GLOBAL    FUNC malloc
[0x00001820]> iS
[Sections]
Nm Paddr       Size Vaddr      Memsz Perms Name
00 0x00000000     0 0x00000000     0 ----
01 0x00000238    28 0x00000238    28 -r-- .interp
02 0x00000254    32 0x00000254    32 -r-- .note.ABI_tag
03 0x00000274    36 0x00000274    36 -r-- .note.gnu.build_id
04 0x00000298    64 0x00000298    64 -r-- .gnu.hash
05 0x000002d8  1656 0x000002d8  1656 -r-- .dynsym
06 0x00000950   756 0x00000950   756 -r-- .dynstr
07 0x00000c44   138 0x00000c44   138 -r-- .gnu.version
08 0x00000cd0    96 0x00000cd0    96 -r-- .gnu.version_r
09 0x00000d30   912 0x00000d30   912 -r-- .rela.dyn
10 0x000010c0  1320 0x000010c0  1320 -r-- .rela.plt
11 0x000015e8    23 0x000015e8    23 -r-x .init
12 0x00001600   896 0x00001600   896 -r-x .plt
13 0x00001980     8 0x00001980     8 -r-x .plt.got
14 0x00001990 16569 0x00001990 16569 -r-x .text
15 0x00005a4c     9 0x00005a4c     9 -r-x .fini
16 0x00005a60  3872 0x00005a60  3872 -r-- .rodata
17 0x00006980   628 0x00006980   628 -r-- .eh_frame_hdr
18 0x00006bf8  3268 0x00006bf8  3268 -r-- .eh_frame
19 0x00007c28     8 0x00207c28     8 -rw- .init_array
20 0x00007c30     8 0x00207c30     8 -rw- .fini_array
21 0x00007c38     8 0x00207c38     8 -rw- .jcr
22 0x00007c40   408 0x00207c40   408 -rw- .data.rel.ro
23 0x00007dd8   480 0x00207dd8   480 -rw- .dynamic
24 0x00007fb8    48 0x00207fb8    48 -rw- .got
25 0x00008000   464 0x00208000   464 -rw- .got.plt
26 0x000081e0   192 0x002081e0   192 -rw- .data
27 0x000082a0     0 0x002082a0   416 -rw- .bss
28 0x000082a0    52 0x00000000    52 ---- .gnu_debuglink
29 0x000082d4   271 0x00000000   271 ---- .shstrtab

```
In `ls` `malloc` goes to the `.got` section instead of the `.plt.got` so it will be patched by the loader at launch time - since the lazy binding process does not take place here - I assume full relro binaries have been tested already.

```
ls
[0x0021efd0]> ii~malloc
  76 0x00000000  GLOBAL    FUNC malloc
[0x0021efd0]> pd 4
            ;-- reloc.malloc:
            0x0021efd0      .qword 0x0000000000000000                  ; RELOC 64 malloc
            ;-- reloc._Jv_RegisterClasses:
            0x0021efd8      .qword 0x0000000000000000                  ; RELOC 64 _Jv_RegisterClasses
            ;-- reloc._ITM_registerTMCloneTable:
            0x0021efe0      .qword 0x0000000000000000                  ; RELOC 64 _ITM_registerTMCloneTable
            ;-- reloc.__cxa_finalize:
            0x0021efe8      .qword 0x0000000000000000                  ; RELOC 64 __cxa_finalize
[0x0021efd0]> iS
[Sections]
Nm Paddr       Size Vaddr      Memsz Perms Name
00 0x00000000     0 0x00000000     0 ----
01 0x00000238    28 0x00000238    28 -r-- .interp
02 0x00000254    32 0x00000254    32 -r-- .note.ABI_tag
03 0x00000274    36 0x00000274    36 -r-- .note.gnu.build_id
04 0x00000298   180 0x00000298   180 -r-- .gnu.hash
05 0x00000350  3264 0x00000350  3264 -r-- .dynsym
06 0x00001010  1501 0x00001010  1501 -r-- .dynstr
07 0x000015ee   272 0x000015ee   272 -r-- .gnu.version
08 0x00001700   112 0x00001700   112 -r-- .gnu.version_r
09 0x00001770  4968 0x00001770  4968 -r-- .rela.dyn
10 0x00002ad8  2544 0x00002ad8  2544 -r-- .rela.plt
11 0x000034c8    23 0x000034c8    23 -r-x .init
12 0x000034e0  1712 0x000034e0  1712 -r-x .plt
13 0x00003b90    32 0x00003b90    32 -r-x .plt.got
14 0x00003bb0 73881 0x00003bb0 73881 -r-x .text
15 0x00015c4c     9 0x00015c4c     9 -r-x .fini
16 0x00015c60 20244 0x00015c60 20244 -r-- .rodata
17 0x0001ab74  2092 0x0001ab74  2092 -r-- .eh_frame_hdr
18 0x0001b3a0 11748 0x0001b3a0 11748 -r-- .eh_frame
19 0x0001e388     8 0x0021e388     8 -rw- .init_array
20 0x0001e390     8 0x0021e390     8 -rw- .fini_array
21 0x0001e398     8 0x0021e398     8 -rw- .jcr
22 0x0001e3a0  2584 0x0021e3a0  2584 -rw- .data.rel.ro
23 0x0001edb8   496 0x0021edb8   496 -rw- .dynamic
24 0x0001efa8    72 0x0021efa8    72 -rw- .got
25 0x0001f000   872 0x0021f000   872 -rw- .got.plt
26 0x0001f380   616 0x0021f380   616 -rw- .data
27 0x0001f5e8     0 0x0021f600  4552 -rw- .bss
28 0x0001f5e8    52 0x00000000    52 ---- .gnu_debuglink
29 0x0001f61c   271 0x00000000   271 ---- .shstrtab
```
Attached `cat` as well to ease your testing.

![cat](https://user-images.githubusercontent.com/3474042/51353710-a322a100-1ab1-11e9-8580-ee0ce31f5090.png)

**Comment by mxz297** [comment]
Closed by via #548
